{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "Coding 1: (20 points)\n",
    "\n",
    "Design your own object class. It should be an object that is useful to solve a problem relevant to\n",
    "your daily life, such as a protein sequence or perhaps a bank account object. Good commenting\n",
    "is a must! Docstrings are not required, however. The object must have the following\n",
    "characteristics:\n",
    "\n",
    "1. A brief written description of the class, its purpose, and how to use it. (5 points)\n",
    "2. A well-designed init, getters, and setters. (5 points)\n",
    "3. At least two extra methods that serve useful purposes. (5 points)\n",
    "4. Design two subclasses of this object that extend it in a useful, meaningful way. (5 points)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "outputs": [],
   "source": [
    "import random\n",
    "import re\n",
    "\n",
    "class Sequence:\n",
    "    \"\"\"This is a representation of any biological sequence. This specific class is\n",
    "    intended to be extended as a base class and is not intended to be implemented\n",
    "    on its own (although it can be). The base Sequence class does not have a type,\n",
    "    although child classes could (e.g. DNA/RNA/Protein sequences). The sequence\n",
    "    itself should not be able to be arbitrarily set, and neither should its type.\n",
    "    It also has a \"quality string\" which denotes the confidence that each element\n",
    "    of the sequence is actually correct. For example, in the case that this sequence\n",
    "    represents a DNA sequence, the first value of the quality string represents\n",
    "    the confidence that the first nucleotide base call is actually correct. By\n",
    "    default, the quality string is empty, but it should always be the same length\n",
    "    as the sequence itself.\n",
    "\n",
    "    A Sequence object can be created by simply passing the sequence as a string to\n",
    "    the constructor method. Sequences can have their quality string set, they can\n",
    "    have single elements mutated (either at random or at a specific index with a\n",
    "    specific value), inserted, or deleted.\n",
    "\n",
    "    Example usage:\n",
    "    # create an amino acid sequence\n",
    "    > aa_sequence = Sequence(\"RLLKGEENWYQIIAAA\")\n",
    "\n",
    "    # return the sequence using the getter method\n",
    "    > aa_sequence.get_sequence()\n",
    "\n",
    "    # mutate a random amino acid residue at a random position\n",
    "    > aa_sequence.mutate()\n",
    "\n",
    "    # mutate a random amino acid residue into a Tryptophan\n",
    "    > aa_sequence.mutate(mutation_value=\"W\")\n",
    "\n",
    "    # delete the last amino acid residue\n",
    "    > aa_sequence.delete_element(15)\n",
    "\n",
    "    :param sequence: the sequence to be represented\n",
    "    :type sequence: str\n",
    "    \"\"\"\n",
    "    def __init__(self, sequence: str):\n",
    "        self.__sequence_string = sequence\n",
    "        self.__sequence_type = None\n",
    "        self.__quality_string = \"\"\n",
    "\n",
    "    def _valid_index(self, index: int):\n",
    "        \"\"\"helper function for checking if index is valid in other member functions\"\"\"\n",
    "        if index < 0 or index > len(self.__sequence_string):\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "    def _valid_sequence(self) -> None:\n",
    "        \"\"\"checking for a valid sequence doesn't matter in the base case\"\"\"\n",
    "        pass\n",
    "\n",
    "    def get_sequence(self) -> str:\n",
    "        \"\"\"getter function to return the sequence\"\"\"\n",
    "        return self.__sequence_string\n",
    "\n",
    "    def get_sequence_length(self) -> int:\n",
    "        \"\"\"getter function to return the length of the sequence\"\"\"\n",
    "        return len(self.__sequence_string)\n",
    "\n",
    "    def get_sequence_type(self) -> str:\n",
    "        \"\"\"getter function to return the sequence type\"\"\"\n",
    "        return self.__sequence_type\n",
    "\n",
    "    def get_quality_values(self) -> str:\n",
    "        \"\"\"getter function to return the quality values\"\"\"\n",
    "        return self.__quality_string\n",
    "\n",
    "    def update_quality_values(self, quality_values) -> None:\n",
    "        \"\"\"setter function to change the sequence to something completely new\"\"\"\n",
    "        if len(quality_values) != self.get_sequence_length():\n",
    "            raise ValueError(\"The quality value sequence should be the same length as the sequence\")\n",
    "        self.__quality_string = quality_values\n",
    "\n",
    "    def mutate(self, index: int=None, mutation_value: str=None) -> None:\n",
    "        \"\"\"\n",
    "        extra method #1: mutate a value at a given index in the sequence\n",
    "        the mutated value will be one of the existing values in the sequence unless\n",
    "        the mutation value is explicitly defined. the mutation may be silent (e.g.\n",
    "        the value will be the same)\n",
    "\n",
    "        :param index: the index to mutate at. sequence is 0-indexed\n",
    "        :param mutation_value: value to mutate at a specified index in the sequence.\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        if index and not self._valid_index(index):\n",
    "            raise IndexError(\"Tried to insert outside of the sequence bounds\")\n",
    "        elif not index:\n",
    "            index = random.choice(range(0, len(self.__sequence_string)))\n",
    "\n",
    "        if not mutation_value:\n",
    "            mutation_value = random.choice(list(set(self.__sequence_string)))\n",
    "\n",
    "        if index == 0:\n",
    "            self.__sequence_string = mutation_value + self.__sequence_string[index + 1:]\n",
    "        else:\n",
    "            self.__sequence_string[:index] + mutation_value + self.__sequence_string[index + 1:]\n",
    "\n",
    "    def insert_element(self, index: int, insertion_value: str) -> None:\n",
    "        \"\"\"\n",
    "        extra method #2: insert a value at a given index in the sequence\n",
    "\n",
    "        :param index: the index to insert an element into the sequence\n",
    "        :param insertion_value: the value to insert at the provided index\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        if not self._valid_index(index):\n",
    "            raise IndexError(\"Tried to insert outside of the sequence bounds\")\n",
    "\n",
    "        self.__sequence_string = self.__sequence_string[:index] + insertion_value + self.__sequence_string[index:]\n",
    "\n",
    "    def delete_element(self, index: int) -> None:\n",
    "        \"\"\"\n",
    "        extra method #3: delete a value at a given index in the sequence\n",
    "\n",
    "        :param index: the index to delete an element in the sequence from\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        if not self._valid_index(index):\n",
    "            raise IndexError(\"Tried to delete outside of the sequence bounds\")\n",
    "\n",
    "        self.__sequence_string = self.__sequence_string[:index] + self.__sequence_string[index+1:]"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "outputs": [],
   "source": [
    "class DNASequence(Sequence):\n",
    "    \"\"\"A representation of a DNA sequence.\n",
    "\n",
    "    In addition to the functionality of the base class, a DNA sequence\n",
    "    can also transcribe an RNA sequence from itself, and can also recall\n",
    "    its complementary sequence.\n",
    "\n",
    "    Example usage:\n",
    "    # create a DNA sequence\n",
    "    > my_dna_sequence = DNASequence(\"ATCCCGGCGATATA\")\n",
    "\n",
    "    # get the complementary strand\n",
    "    > my_dna_sequence.complement()\n",
    "\n",
    "    # get the RNA transcription of this strand\n",
    "    > my_dna_sequence.transcribe()\n",
    "\n",
    "    :param sequence: the DNA sequence\n",
    "    :type sequence: str\n",
    "    \"\"\"\n",
    "    def __init__(self, sequence: str):\n",
    "        super().__init__(sequence)\n",
    "        if not self._valid_sequence():\n",
    "            raise ValueError(\"Not a valid DNA sequence, double check the input bases\")\n",
    "        self.__sequence_type = \"DNA\"\n",
    "\n",
    "    def _valid_sequence(self) -> bool:\n",
    "        \"\"\"\n",
    "        check if the sequence is a valid DNA sequence\n",
    "\n",
    "        :return: whether the sequence is a valid DNA sequence\n",
    "        \"\"\"\n",
    "        if any([base.lower() not in {\"a\", \"t\", \"g\", \"c\"} for base in self.get_sequence()]):\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "    def complement(self) -> str:\n",
    "        \"\"\"\n",
    "        return the complementary DNA sequence\n",
    "\n",
    "        :return: the complementary/reverse DNA strand\n",
    "        \"\"\"\n",
    "        nucleotide_map = {\n",
    "            \"A\": \"T\",\n",
    "            \"T\": \"A\",\n",
    "            \"G\": \"C\",\n",
    "            \"C\": \"G\",\n",
    "        }\n",
    "        reverse_strand = \"\".join([nucleotide_map[base.upper()] for base in self.get_sequence()])\n",
    "\n",
    "        if self.get_sequence()[0].isupper():\n",
    "            return reverse_strand\n",
    "        else:\n",
    "            return reverse_strand.lower()\n",
    "\n",
    "    def transcribe(self) -> str:\n",
    "        \"\"\"\n",
    "        return the RNA sequence that would be transcribed from this sequence\n",
    "\n",
    "        :return: the RNA version of the sequence\n",
    "        \"\"\"\n",
    "        if self.get_sequence()[0].isupper():\n",
    "            return re.sub(r\"T\", \"U\", self.get_sequence())\n",
    "        else:\n",
    "            return re.sub(r\"t\", \"u\", self.get_sequence())"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "outputs": [],
   "source": [
    "class RNASequence(Sequence):\n",
    "    \"\"\"A representation of an RNA sequence\n",
    "\n",
    "    In addition to the functionality of the base class, an RNA sequence\n",
    "    can also translate an amino acid sequence from itself\n",
    "\n",
    "    Example usage:\n",
    "    # create an RNA sequence\n",
    "    > my_rna_sequence = RNASequence(\"AUGAAGGCUCGGAUAUAGGG\")\n",
    "\n",
    "    # translate the RNA sequence into an amino acid sequence\n",
    "    > my_rna_sequence.translate()\n",
    "\n",
    "    :param sequence: the RNA sequence\n",
    "    :type sequence: str\n",
    "    \"\"\"\n",
    "    def __init__(self, sequence):\n",
    "        super().__init__(sequence)\n",
    "        if not self._valid_sequence():\n",
    "            raise ValueError(\"Not a valid RNA sequence, double check the input bases\")\n",
    "        self.__sequence_type = \"RNA\"\n",
    "\n",
    "    def _valid_sequence(self) -> bool:\n",
    "        \"\"\"\n",
    "        check if the sequence is a valid DNA sequence\n",
    "\n",
    "        :return: whether the sequence is a valid DNA sequence\n",
    "        \"\"\"\n",
    "        if any([base.lower() not in {\"a\", \"u\", \"g\", \"c\"} for base in self.get_sequence()]):\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "    def _generate_codon(self) -> str:\n",
    "        \"\"\"\n",
    "        helper function to split the RNA sequence into codons\n",
    "\n",
    "        :return: a 3-base codon\n",
    "        \"\"\"\n",
    "        remaining_bases = self.get_sequence()\n",
    "\n",
    "        while remaining_bases:\n",
    "            yield remaining_bases[:3]\n",
    "            remaining_bases = remaining_bases[3:]\n",
    "\n",
    "    def translate(self) -> str:\n",
    "        \"\"\"\n",
    "        translate the RNA sequence into a sequence of amino acids. the length of the sequence\n",
    "        should be divisible by 3\n",
    "\n",
    "        :return: the translated amino acid sequence from the RNA sequence\n",
    "        \"\"\"\n",
    "        if self.get_sequence_length() % 3 != 0:\n",
    "            raise ValueError(\"The length of the RNA sequence is not divisible by 3, not sure which reading frame to use\")\n",
    "\n",
    "        codon_dict = {\"UUU\": \"F\", \"UUC\": \"F\",\n",
    "                      \"UUA\": \"L\", \"UUG\": \"L\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "                      \"AUU\": \"I\", \"AUA\": \"I\", \"AUC\": \"I\",\n",
    "                      \"AUG\": \"M\",\n",
    "                      \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "                      \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"AGU\": \"S\", \"AGC\": \"S\",\n",
    "                      \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "                      \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "                      \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "                      \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
    "                      \"CAU\": \"H\", \"CAC\": \"H\",\n",
    "                      \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "                      \"AAU\": \"N\", \"AAC\": \"N\",\n",
    "                      \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "                      \"GAU\": \"D\", \"GAC\": \"D\",\n",
    "                      \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "                      \"UGU\": \"C\", \"UGC\": \"C\",\n",
    "                      \"UGG\": \"W\",\n",
    "                      \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "                      \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "                      \"UAA\": \"[STOP]\", \"UAG\": \"[STOP]\", \"UGA\": \"[STOP]\",\n",
    "        }\n",
    "\n",
    "        aa_chain = \"\"\n",
    "        for codon in self._generate_codon():\n",
    "            aa_chain += codon_dict[codon]\n",
    "\n",
    "        return aa_chain\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Coding 2: (20 points)\n",
    "\n",
    "Incomprehensible!\n",
    "\n",
    "Create two comprehensions of any kind you like. For each one, do the following (10 points, 1\n",
    "per comprehension, 2 for each of the following):\n",
    "\n",
    "1. Attempt to create the comprehension as a for loop. If it is impossible, say why.\n",
    "2. Attempt to create the comprehension as a lambda (with any friends like\n",
    "map/filter/reduce). If it is impossible, say why."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "outputs": [],
   "source": [
    "# a list comprehension -> a for loop\n",
    "daisy_comprehension = [\"she loves me\" if num % 2 else \"she loves me not\" for num in range(30)]\n",
    "\n",
    "daisy_loop = []\n",
    "for num in range(30):\n",
    "    if num % 2:\n",
    "        daisy_loop.append(\"she loves me\")\n",
    "    else:\n",
    "        daisy_loop.append(\"she loves me not\")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "outputs": [],
   "source": [
    "# a dict comprehension -> a lambda + map\n",
    "names = [\"Pat\", \"Ron\", \"Mo\", \"Nehemiah\", \"Omerit\", \"Marvin\", \"Isaiah\"]\n",
    "who_let_the_dogs_out_comprehension = {name: (True if name == \"Isaiah\" else False) for name in names}\n",
    "\n",
    "who_let_the_dogs_out_lambda = lambda x: x == \"Isaiah\"\n",
    "\n",
    "who_let_the_dogs_out_map = {name: next(map(who_let_the_dogs_out_lambda, [name])) for name in names}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Coding 3: (10 points)\n",
    "\n",
    "Efficiency exploration: we’re advanced now!\n",
    "\n",
    "Devise two pieces of code that are sufficiently unique from any other code you wrote in the\n",
    "prior questions. They should be loops of some kind that perform a finite task (no while True\n",
    "stuff!), even if that task is silly.  Include a brief description of the purpose of each piece of code.\n",
    "(1 point each)\n",
    "\n",
    "Time their execution using the timeit module as here:\n",
    "https://docs.python.org/3.8/library/timeit.html (1 point each)\n",
    "\n",
    "Try making some change in your code that you think might make the code more efficient but not\n",
    "alter its function - that is, it must produce the same result as its unaltered form. If you do, show\n",
    "the change and record your difference in time. If you are unable to find an adjustment that does\n",
    "so, mention what you tried. (2 points each)\n",
    "\n",
    "Please include a viable explanation of why your change made the code run more quickly, or\n",
    "why the change failed to do so. (1 point each)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [],
   "source": [
    "import timeit"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5000000050000000\n",
      "For-loop time: 3.5041889999993145\n",
      "-----\n",
      "5000000050000000\n",
      "sum func time: 1.0773054159944877\n"
     ]
    }
   ],
   "source": [
    "# code snippet 1: compare ways of finding the sum of 0 to 100000000\n",
    "\n",
    "def for_loop_sum() -> None:\n",
    "    \"\"\"\n",
    "    slowest way to calculate; use a naive for-loop to\n",
    "    loop over integers from 0 to 100000000\n",
    "\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    iterations = 100000000\n",
    "    counter = 0\n",
    "    for i in range(iterations + 1):\n",
    "        counter += i\n",
    "    print(counter)\n",
    "\n",
    "def sum_sum() -> None:\n",
    "    \"\"\"\n",
    "    faster way to calculate; use the built-in `sum`\n",
    "    function to sum the list of integers more quickly.\n",
    "\n",
    "    the increase in speed is likely due to `sum` being implemented\n",
    "    and rigorously optimized in C inside the Python interpreter,\n",
    "    whereas the for-loop has to be interpreted and is thus slower.\n",
    "\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    print(sum(range(100000000 + 1)))\n",
    "\n",
    "print(f'For-loop time: {timeit.timeit(\"for_loop_sum()\", number=1, setup=\"from __main__ import for_loop_sum\")}')\n",
    "print(\"-----\")\n",
    "print(f'sum func time: {timeit.timeit(\"sum_sum()\", number=1, setup=\"from __main__ import sum_sum\")}')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For-loop time: 2.7473626249702647\n",
      "-----\n",
      "Comprehension time: 2.3807997920084745\n"
     ]
    }
   ],
   "source": [
    "# code snippet 2: appending to a list\n",
    "\n",
    "def for_loop_append():\n",
    "    \"\"\"\n",
    "    slow way to create a list of all integers from 0 to 100000000\n",
    "\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    list_to_append = []\n",
    "    iterations = 100000000\n",
    "\n",
    "    for i in range(iterations + 1):\n",
    "        list_to_append.append(i)\n",
    "\n",
    "def comprehension_append():\n",
    "    \"\"\"\n",
    "    faster way to create a list of all integers from 0 to 100000000\n",
    "\n",
    "    this approach is might be faster if the list is created wholesale at the start of the\n",
    "    comprehension call. since memory is already allocated, filling the list is a trivial task.\n",
    "    in contrast, the for-loop implementation needs to expand how much memory is allocated to\n",
    "    the list with each iteration, which may increase runtime. the for-loop implementation\n",
    "    also needs to look up the `.append()` method with each iteration, which may require\n",
    "    additional runtime.\n",
    "\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    iterations = 100000000\n",
    "    list_comprehension_append = [i for i in range(iterations + 1)]\n",
    "\n",
    "print(f'For-loop time: {timeit.timeit(\"for_loop_append()\", number=1, setup=\"from __main__ import for_loop_append\")}')\n",
    "print(\"-----\")\n",
    "print(f'Comprehension time: {timeit.timeit(\"comprehension_append()\", number=1, setup=\"from __main__ import comprehension_append\")}')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}